src/system/SystemPacketBuffer.cpp

```diff
@@ -207,25 +208,36 @@ void PacketBuffer::SetStart(uint8_t * aNewStart)
         aNewStart = kEnd;
 
     ptrdiff_t lDelta = aNewStart - static_cast<uint8_t *>(this->payload);
-    if (lDelta > this->len)
-        lDelta = this->len;
+    if (lDelta > static_cast<int32_t>(this->len))
+        lDelta = static_cast<int32_t>(this->len);
 
-    this->len     = static_cast<uint16_t>(static_cast<ptrdiff_t>(this->len) - lDelta);
-    this->tot_len = static_cast<uint16_t>(static_cast<ptrdiff_t>(this->tot_len) - lDelta);
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    this->len     = static_cast<uint16_t>(static_cast<int32_t>(this->len) - lDelta);
+    this->tot_len = static_cast<uint16_t>(static_cast<int32_t>(this->tot_len) - lDelta);
+#else
+    this->len     = static_cast<size_t>(static_cast<int32_t>(this->len) - lDelta);
+    this->tot_len = static_cast<size_t>(static_cast<int32_t>(this->tot_len) - lDelta);
+#endif
     this->payload = aNewStart;
 }
 
-void PacketBuffer::SetDataLength(uint16_t aNewLen, PacketBuffer * aChainHead)
+void PacketBuffer::SetDataLength(size_t aNewLen, PacketBuffer * aChainHead)
 {
-    const uint16_t kMaxDataLen = this->MaxDataLength();
+    const size_t kMaxDataLen = this->MaxDataLength();
 
     if (aNewLen > kMaxDataLen)
         aNewLen = kMaxDataLen;
 
-    ptrdiff_t lDelta = static_cast<ptrdiff_t>(aNewLen) - static_cast<ptrdiff_t>(this->len);
+    int32_t lDelta = static_cast<int32_t>(aNewLen) - static_cast<int32_t>(this->len);
```

Comment:
Why is this not using ptrdiff_t?  That seems like the right type to use.

Reply:
ptrdiff_t is not easily inter-castable with size_t on some platforms, IIRC. Hence the change.

Reply:
This is why ssize_t exists, no?